### ISSUE TITLE
Brief garbage collection explanation in Chapter 3 on closures.

### ISSUE THREAD

**Issue by arsen91** [getify/You-Dont-Know-JS#1456]
Hello. Thanks for the material - very extensive and helpful!

Question about GC brief explanation in chapter 3 on [closures](https://github.com/getify/You-Dont-Know-JS/blob/master/scope%20%26%20closures/ch3.md#garbage-collection).

You are writing:

> The `click` function click handler callback doesn't need the someReallyBigData variable at all. That means, theoretically, after process(..) runs, the big memory-heavy data structure could be garbage collected. However, it's quite likely (though implementation dependent) that the JS engine will still have to keep the structure around, since the click function has a closure over the entire scope.

However, according to the [docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Memory_Management#References) GC will NOT collect anything only in one scenario - if there is no reference. Otherwise, GC will collect the object. In your example closure does not have a reference.

In your example, although it's not explicitly said, it's assumed (at least by me) that the code is written in the global scope. If it is the case, this `someReallyBigData` object will never be collected, it will always remain in the global scope which never ends. Only if we apply your fix - block+let will it delete the object. We also can wrap it in a function, but it does not depend on the fact that we have a callback for a click, which creates a closure (which we're not using in the example). 

To make this example purer I put everything in IIFE. This will let GC clean up everything after IIFE is done executing. Following your logic in the explanation, I would expect in this [snippet](https://gist.github.com/arsen91/aaba5d7c65ce437eea38da11507ace21) that GC will keep `dates` array in case the callback needs it. However, when I open it in the browser and inspect Memory tab in Google Dev Tools, I don't see this big array of dates in the list of stored objects in memory snapshot.

There is one thing I can do though, to make GC skip collecting this big object - put the `console.log(dates)` in the click callback. This way we'll have a reference to the array and GC won't collect it. I confirmed it taking one more memory snapshot.

Although your fix with let+block works, it is kind of misleading to say that because there is a closure created by click callback and it can potentially need the object GC won't delete it. From my understanding, if you explicitly do not reference the object in the closure GC **will** collect it. 

I also see that you're using words like `theoretically` and `quite likely`, but per the documentation and actual tests I made, it's pretty straightforward - if there is no reference GC collects, if there is a reference - GC preserves the object.

Please let me know if everything is clear, if I need to give more details or if I misunderstood the example and should close the ticket as not relevant continue my learning journey. 

Thank you!

**Comment by getify** [comment]
> if you explicitly do not reference the object in the closure GC will collect it.

This is an implementation detail. It's not guaranteed that it will, or won't, be GC'd. I have observed things still in closure even in the case where it would seem GC should have cleaned it up. 

For example, what if there's an eval(..) in the inner function.. can JS really guarantee that you won't use  it to make a dynamic reference to the big array? Same concern with the `with` statement. At least at some point in the recent past, I have seen JS engines being conservative with GC for these sorts of reasons.

It continues to be my advice that people assume that the entire scope is closed over, not each individual variable/value.

**Comment by arsen91** [comment]
I've tried to put an empty `eval("")` in the callback of click and then, as you correctly stated earlier in the book - js engine doesn't do the proper performance optimizations, thus dates array **is** in the memory and not GC'ed. 

However, if there is no reference to the array or there is no use of `eval()` or `with` in the function that creates closure - `dates` is GC'ed in the browser. I don't have enough time to go deeper in memory debugging to check out how nodejs does that. Maybe some other engines are doing it differently, but I'm pretty sure that in the "average" environments like browser, node it will be that simple - no reference - optimized, there is a reference to an object or eval or with - not optimized to make sure GC doesn't delete potentially important stuff.

**Comment by getify** [comment]
As I said, this is implementation dependent. It used to not do this GC. Just because it is now, under certain circumstances, doesn't mean it always will be the case. My advice continues to be, be conservative in those assumptions.

**Comment by arsen91** [comment]
My point is not that the advice is not correct. I agree with the advice to be conservative with closures.

I feel misleading statement and example which states that GC will not collect the reference if there is a callback which creates the closure but has no reference to an object. In the example DOM manipulation is used to create a closure, thus assuming the code is run in the browser. The official documentation says that in the browser (V8) it will collect any object if there's no reference to it regardless of created closure.

I thought explanation can be much cleaner on that and making an explicit note about GC in browsers and current workings would be fair to readers, of course providing, in my opinion, valuable advice about conservatism in this regard. Especially it would make sense when it corresponds to the official documentation. But that is up to the author to be aligned with the docs or not in this particular example. I haven't yet got to detailed GC explanation in the book. I'm sure it's precise and straightforward.

Thank you for your responses and all the best!

**Comment by getify** [comment]
This is what the passage says:

> it's quite likely (though implementation dependent)

That's consistent with what I'm saying here. It's implementation dependent. It depends. Sometimes GC could happen. Other times it might not. Furthermore, at the time I wrote it, no engines that I tested *did* do the GC as you're currently observing. That is proof that what I said is accurate: it's implementation dependent. The spec does not require it.
